### ISSUE TITLE
Change how zero_grad() behaves.

### ISSUE THREAD

**Issue by danieltudosiu** [pytorch/pytorch#42748]
## 🚀 Feature
Please change how zero_grad behave [by default](https://pytorch.org/docs/stable/_modules/torch/optim/optimizer.html#Optimizer.zero_grad) to be like Apex [does it](https://nvidia.github.io/apex/_modules/apex/optimizers/fused_adam.html#FusedAdam.zero_grad).

## Motivation
It signals better to the garbage collector to come and pick up the released tensors. 

## Pitch
For 3D losses such as [Adaptive Loss](https://github.com/jonbarron/robust_loss_pytorch) which release a lot of tensors in the process of calculating the actual loss function the current way results in a behaviour similar to memory leak and it is hard to track if you do not have experience with the Apex package. 

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by ngimel** [comment]
related #41696.

**Comment by albanD** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/issues/41696 that has more discussion a a corresponding PR.
